hydro_lang/src/location/mod.rs

```diff
@@ -11,108 +26,73 @@ use stageleft::{QuotedWithContext, q, quote_type};
 use crate::live_collections::singleton::Singleton;
 use crate::live_collections::stream::{ExactlyOnce, NoOrder, Stream, TotalOrder};
 use crate::location::cluster::ClusterIds;
+use crate::location::dynamic::LocationId;
 use crate::location::external_process::{
     ExternalBincodeBidi, ExternalBincodeSink, ExternalBytesPort, Many,
 };
 use crate::nondet::{NonDet, nondet};
 use crate::staging_util::get_this_crate;
 
+pub mod dynamic;
+
+#[expect(missing_docs, reason = "TODO")]
 pub mod external_process;
 pub use external_process::External;
 
+#[expect(missing_docs, reason = "TODO")]
 pub mod process;
 pub use process::Process;
 
+#[expect(missing_docs, reason = "TODO")]
 pub mod cluster;
 pub use cluster::Cluster;
 
+#[expect(missing_docs, reason = "TODO")]
 pub mod member_id;
 pub use member_id::MemberId;
 
+#[expect(missing_docs, reason = "TODO")]
 pub mod tick;
 pub use tick::{Atomic, NoTick, Tick};
 
-#[derive(PartialEq, Eq, Clone, Debug, Hash, Serialize, Deserialize)]
-pub enum LocationId {
-    Process(usize),
-    Cluster(usize),
-    Tick(usize, Box<LocationId>),
-}
-
+#[expect(missing_docs, reason = "TODO")]
 #[derive(PartialEq, Eq, Clone, Debug, Hash, Serialize, Deserialize)]
 pub enum MembershipEvent {
     Joined,
     Left,
 }
 
-impl LocationId {
-    pub fn root(&self) -> &LocationId {
-        match self {
-            LocationId::Process(_) => self,
-            LocationId::Cluster(_) => self,
-            LocationId::Tick(_, id) => id.root(),
-        }
-    }
-
-    pub fn is_root(&self) -> bool {
-        match self {
-            LocationId::Process(_) | LocationId::Cluster(_) => true,
-            LocationId::Tick(_, _) => false,
-        }
-    }
-
-    pub fn raw_id(&self) -> usize {
-        match self {
-            LocationId::Process(id) => *id,
-            LocationId::Cluster(id) => *id,
-            LocationId::Tick(_, _) => panic!("cannot get raw id for tick"),
-        }
-    }
-
-    pub fn swap_root(&mut self, new_root: LocationId) {
-        match self {
-            LocationId::Tick(_, id) => {
-                id.swap_root(new_root);
-            }
-            _ => {
-                assert!(new_root.is_root());
-                *self = new_root;
-            }
-        }
-    }
-}
-
-pub fn check_matching_location<'a, L: Location<'a>>(l1: &L, l2: &L) {
-    assert_eq!(l1.id(), l2.id(), "locations do not match");
-}
-
+#[expect(missing_docs, reason = "TODO")]
 #[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
 pub enum NetworkHint {
     Auto,
     TcpPort(Option<u16>),
 }
 
-pub trait Location<'a>: Clone {
+pub(crate) fn check_matching_location<'a, L: Location<'a>>(l1: &L, l2: &L) {
+    assert_eq!(Location::id(l1), Location::id(l2), "locations do not match");
+}
+
+#[expect(missing_docs, reason = "TODO")]
+#[expect(
+    private_bounds,
+    reason = "only internal Hydro code can define location types"
+)]
+pub trait Location<'a>: dynamic::DynLocation {
```

Comment:
I guess `private_bounds` implies it as well, but would be nice to do `#[sealed::sealed]` on these and its impls